Test Failure (20180730-0122): JDBCDriverManagerTest.testBasicConnection

Test Failure (20180730-0122): com.ibm.ws.jdbc.fat.driver.JDBCDriverManagerTest.testBasicConnection

junit.framework.AssertionFailedError: 2018-07-30-09:12:05:028 The response did not contain "[SUCCESS]".  Full output is:"
ERROR: Caught exception attempting to call test method testBasicConnection on servlet jdbc.fat.driver.web.JDBCDriverManagerServlet
java.sql.SQLException
	at com.ibm.ws.rsadapter.AdapterUtil.toSQLException(AdapterUtil.java:822)
	at com.ibm.ws.rsadapter.jdbc.WSJdbcDataSource.getConnection(WSJdbcDataSource.java:152)
	at com.ibm.ws.rsadapter.jdbc.WSJdbcDataSource.getConnection(WSJdbcDataSource.java:116)
	at jdbc.fat.driver.web.JDBCDriverManagerServlet.testBasicConnection(JDBCDriverManagerServlet.java:53)
	at componenttest.app.FATServlet.doGet(FATServlet.java:71)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:687)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at com.ibm.ws.webcontainer.servlet.ServletWrapper.service(ServletWrapper.java:1255)
	at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:743)
	at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:440)
	at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1208)
	at com.ibm.ws.webcontainer.webapp.WebApp.handleRequest(WebApp.java:4954)
	at com.ibm.ws.webcontainer.osgi.DynamicVirtualHost$2.handleRequest(DynamicVirtualHost.java:314)
	at com.ibm.ws.webcontainer.WebContainer.handleRequest(WebContainer.java:996)
	at com.ibm.ws.webcontainer.osgi.DynamicVirtualHost$2.run(DynamicVirtualHost.java:279)
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink$TaskWrapper.run(HttpDispatcherLink.java:1011)
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.wrapHandlerAndExecute(HttpDispatcherLink.java:414)
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.ready(HttpDispatcherLink.java:373)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleDiscrimination(HttpInboundLink.java:532)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleNewRequest(HttpInboundLink.java:466)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.processRequest(HttpInboundLink.java:331)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.ready(HttpInboundLink.java:302)
	at com.ibm.ws.tcpchannel.internal.NewConnectionInitialReadCallback.sendToDiscriminators(NewConnectionInitialReadCallback.java:165)
	at com.ibm.ws.tcpchannel.internal.NewConnectionInitialReadCallback.complete(NewConnectionInitialReadCallback.java:74)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager.requestComplete(WorkQueueManager.java:501)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager.attemptIO(WorkQueueManager.java:571)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager.workerRun(WorkQueueManager.java:926)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager$Worker.run(WorkQueueManager.java:1015)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1157)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:627)
	at java.lang.Thread.run(Thread.java:809)
Caused by: javax.resource.spi.ResourceAllocationException
	at com.ibm.ejs.j2c.FreePool.createManagedConnectionWithMCWrapper(FreePool.java:1579)
	at com.ibm.ejs.j2c.FreePool.createOrWaitForConnection(FreePool.java:1231)
	at com.ibm.ejs.j2c.PoolManager.reserve(PoolManager.java:1448)
	at com.ibm.ejs.j2c.ConnectionManager.allocateMCWrapper(ConnectionManager.java:573)
	at com.ibm.ejs.j2c.ConnectionManager.allocateConnection(ConnectionManager.java:309)
	at com.ibm.ws.rsadapter.jdbc.WSJdbcDataSource.getConnection(WSJdbcDataSource.java:143)
Caused by: java.lang.NullPointerException
	at java.util.Hashtable.containsKey(Hashtable.java:458)
	at sun.jdbc.odbc.ee.ConnectionPoolFactory.obtainConnectionPool(ConnectionPoolFactory.java:45)
	at sun.jdbc.odbc.ee.ConnectionPoolDataSource.getPool(ConnectionPoolDataSource.java:433)
	at sun.jdbc.odbc.ee.ConnectionPoolDataSource.getPooledConnection(ConnectionPoolDataSource.java:159)
	at com.ibm.ws.rsadapter.impl.DatabaseHelper$1.run(DatabaseHelper.java:949)
	at java.security.AccessController.doPrivileged(AccessController.java:452)
	at com.ibm.ws.rsadapter.impl.DatabaseHelper.getPooledConnection(DatabaseHelper.java:958)
	at com.ibm.ws.rsadapter.impl.WSManagedConnectionFactoryImpl.getConnection(WSManagedConnectionFactoryImpl.java:813)
	at com.ibm.ws.rsadapter.impl.WSManagedConnectionFactoryImpl.createManagedConnection(WSManagedConnectionFactoryImpl.java:659)
	at com.ibm.ejs.j2c.FreePool.createManagedConnectionWithMCWrapper(FreePool.java:1357)
". expected:<0> but was:<1>
	at componenttest.topology.utils.HttpUtils.findStringInHttpConnection(HttpUtils.java:537)
	at componenttest.topology.utils.HttpUtils.findStringInHttpConnection(HttpUtils.java:495)
	at componenttest.topology.utils.HttpUtils.findStringInUrl(HttpUtils.java:472)
	at componenttest.topology.utils.HttpUtils.findStringInReadyUrl(HttpUtils.java:444)
	at componenttest.topology.utils.HttpUtils.findStringInReadyUrl(HttpUtils.java:414)
	at componenttest.topology.utils.FATServletClient.runTest(FATServletClient.java:81)
	at componenttest.custom.junit.runner.SyntheticServletTest.invokeExplosively(SyntheticServletTest.java:40)
	at componenttest.custom.junit.runner.FATRunner$1.evaluate(FATRunner.java:193)
	at componenttest.custom.junit.runner.FATRunner$2.evaluate(FATRunner.java:323)
	at componenttest.custom.junit.runner.FATRunner.run(FATRunner.java:167)

--- a/dev/com.ibm.ws.jdbc/src/com/ibm/ws/jdbc/internal/JDBCDriverService.java
+++ b/dev/com.ibm.ws.jdbc/src/com/ibm/ws/jdbc/internal/JDBCDriverService.java
@@ -29,6 +29,7 @@ import java.util.Hashtable;
 import java.util.Iterator;
 import java.util.LinkedList;
 import java.util.List;
+import java.util.Map;
 import java.util.Observable;
 import java.util.Properties;
 import java.util.ServiceLoader;
@@ -44,6 +45,7 @@ import javax.sql.XADataSource;
 
 import org.osgi.service.component.ComponentContext;
 
+import com.ibm.websphere.crypto.InvalidPasswordDecodingException;
 import com.ibm.websphere.crypto.PasswordUtil;
 import com.ibm.websphere.ras.Tr;
 import com.ibm.websphere.ras.TraceComponent;
@@ -369,13 +371,21 @@ public class JDBCDriverService extends Observable implements LibraryChangeListen
 
              || null != (className = (String) properties.get(XADataSource.class.getName()))
              || null != (className = JDBCDrivers.getXADataSourceClassName(vendorPropertiesPID))
-             || null != (className = JDBCDrivers.getXADataSourceClassName(getClasspath(sharedLib, true)))
+             || null != (className = JDBCDrivers.getXADataSourceClassName(getClasspath(sharedLib, true))))
+                return create(className, props);
 
-             || nonshipFunction && null != (className = JDBCDrivers.inferDataSourceClassFromDriver(classloader,
-                                                                                JDBCDrivers.CONNECTION_POOL_DATA_SOURCE,
-                                                                                JDBCDrivers.DATA_SOURCE,
-                                                                                JDBCDrivers.XA_DATA_SOURCE)))
-                 return create(className, props);
+            if (nonshipFunction) {
+                Driver driver = loadDriver(props, classloader);
+                if (driver != null)
+                    return driver;
+
+                className = JDBCDrivers.inferDataSourceClassFromDriver(classloader,
+                                                                       JDBCDrivers.CONNECTION_POOL_DATA_SOURCE,
+                                                                       JDBCDrivers.DATA_SOURCE,
+                                                                       JDBCDrivers.XA_DATA_SOURCE);
+                if (className != null)
+                    return create(className, props);
+            }
 
             throw classNotFound(null);
         } finally {
@@ -429,13 +439,21 @@ public class JDBCDriverService extends Observable implements LibraryChangeListen
 
              || null != (className = (String) properties.get(DataSource.class.getName()))
              || null != (className = JDBCDrivers.getDataSourceClassName(vendorPropertiesPID))
-             || null != (className = JDBCDrivers.getDataSourceClassName(getClasspath(sharedLib, true)))
-
-             || nonshipFunction && null != (className = JDBCDrivers.inferDataSourceClassFromDriver(classloader,
-                                                                                JDBCDrivers.XA_DATA_SOURCE,
-                                                                                JDBCDrivers.CONNECTION_POOL_DATA_SOURCE,
-                                                                                JDBCDrivers.DATA_SOURCE)))
+             || null != (className = JDBCDrivers.getDataSourceClassName(getClasspath(sharedLib, true))))
                 return create(className, props);
+
+            if (nonshipFunction) {
+                Driver driver = loadDriver(props, classloader);
+                if (driver != null)
+                    return driver;
+
+                className = JDBCDrivers.inferDataSourceClassFromDriver(classloader,
+                                                                       JDBCDrivers.XA_DATA_SOURCE,
+                                                                       JDBCDrivers.CONNECTION_POOL_DATA_SOURCE,
+                                                                       JDBCDrivers.DATA_SOURCE);
+                if (className != null)
+                    return create(className, props);
+            }
 
             throw classNotFound(null);
         } finally {
@@ -645,6 +663,81 @@ public class JDBCDriverService extends Observable implements LibraryChangeListen
         if (trace && tc.isEntryEnabled())
             Tr.exit(this, tc, "getClasspath", classpath);
         return classpath;
+    }
+
+    /**
+     * Load java.sql.Driver implementations available to the specific class loader and return
+     * the first that accepts the URL (if any) that is specified in the vendor properties.
+     *
+     * @param vProps configured JDBC vendor properties.
+     * @param classloader class loader from which to load JDBC drivers.
+     * @return Driver instance that accepts the URL. NULL if no such Driver can be loaded.
+     * @throws SQLException if an error occurs.
+     */
+    private Driver loadDriver(Properties vProps, ClassLoader classloader) throws SQLException {
+        String url = vProps.getProperty("URL", vProps.getProperty("url"));
+
+        final boolean trace = TraceComponent.isAnyTracingEnabled();
+        if (trace && tc.isDebugEnabled())
+            Tr.entry(this, tc, "loadDriver", url, classloader);
+
+        SQLException failure = null;
+        if (url != null)
+            for (Driver driver : ServiceLoader.load(Driver.class, classloader)) {
+                boolean acceptsURL;
+                try {
+                    acceptsURL = driver.acceptsURL(url);
+                } catch (SQLException x) {
+                    if (failure != null)
+                        failure = x;
+                    acceptsURL = false;
+                }
+                if (acceptsURL) {
+                    // TODO When it does matching, DataSourceService.modified method will dislike that we have stringified the values. Need to allow for this.
+                    // convert property values to String and decode passwords 
+                    for (Map.Entry<Object, Object> prop : vProps.entrySet()) {
+                        Object value = prop.getValue();
+                        if (value instanceof String) {
+                            String str = (String) value;
+                            // Decode passwords
+                            if (PropertyService.isPassword((String) prop.getKey()) && PasswordUtil.getCryptoAlgorithm(str) != null)
+                                try {
+                                    prop.setValue(PasswordUtil.decode(str));
+                                } catch (Exception x) {
+                                    if (trace && tc.isEntryEnabled())
+                                        Tr.exit(this, tc, "loadDriver", x);
+                                    if (x instanceof SQLException)
+                                        throw (SQLException) x;
+                                    else
+                                        throw new SQLNonTransientException(x);
+                                }
+                        } else {
+                            // Convert to String value
+                            prop.setValue(value.toString());
+                        }
+                    }
+
+                    if (classloader != null && url.startsWith("jdbc:derby:") && isDerbyEmbedded.compareAndSet(false, true)) {
+                        embDerbyRefCount.add(classloader);
+                        if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled())
+                            Tr.debug(this, tc, "ref count for shutdown", classloader, embDerbyRefCount);
+                    }
+
+                    if (trace && tc.isEntryEnabled())
+                        Tr.exit(this, tc, "loadDriver", driver);
+                    return driver;
+                } else {
+                    if (trace && tc.isDebugEnabled())
+                        Tr.debug(this, tc, driver + " does not accept url");
+                }
+            }
+
+        if (trace && tc.isEntryEnabled())
+            Tr.exit(this, tc, "loadDriver", failure);
+        if (failure == null)
+            return null;
+        else
+            throw failure;
     }
 
     /**

--- a/dev/com.ibm.ws.jdbc/src/com/ibm/ws/jdbc/internal/JDBCDriverService.java
+++ b/dev/com.ibm.ws.jdbc/src/com/ibm/ws/jdbc/internal/JDBCDriverService.java
@@ -688,7 +688,7 @@ public class JDBCDriverService extends Observable implements LibraryChangeListen
                 try {
                     acceptsURL = driver.acceptsURL(url);
                 } catch (SQLException x) {
-                    if (failure != null)
+                    if (failure == null)
                         failure = x;
                     acceptsURL = false;
                 }

--- a/dev/com.ibm.ws.jdbc/src/com/ibm/ws/jdbc/internal/JDBCDrivers.java
+++ b/dev/com.ibm.ws.jdbc/src/com/ibm/ws/jdbc/internal/JDBCDrivers.java
@@ -363,7 +363,20 @@ public class JDBCDrivers {
         // Load JDBC driver class from service registry and use the package to infer the data source class
         ServiceLoader<Driver> serviceLoader = loader == null ? ServiceLoader.load(Driver.class) : ServiceLoader.load(Driver.class, loader);
         if (serviceLoader != null) {
-            for (Iterator<Driver> it = serviceLoader.iterator(); found[preferredType] == null && it.hasNext(); ) {
+            // In order to give preference to JDBC drivers supplied by the user,
+            // move to the end of the list any JDBC drivers that are packaged with Java
+            List<Driver> drivers = new ArrayList<Driver>();
+            List<Driver> builtInDrivers = new ArrayList<Driver>();
+            for (Iterator<Driver> it = serviceLoader.iterator(); it.hasNext(); ) {
+                Driver driver = it.next();
+                if (driver.getClass().getName().startsWith("sun."))
+                    builtInDrivers.add(driver);
+                else
+                    drivers.add(driver);
+            }
+            drivers.addAll(builtInDrivers);
+
+            for (Iterator<Driver> it = drivers.iterator(); found[preferredType] == null && it.hasNext(); ) {
                 Driver driver = it.next();
                 String driverClassName = driver.getClass().getName();
                 String driverPackage = null;

--- a/dev/com.ibm.ws.jdbc_fat_driver/publish/servers/com.ibm.ws.jdbc.fat.driver/server.xml
+++ b/dev/com.ibm.ws.jdbc_fat_driver/publish/servers/com.ibm.ws.jdbc.fat.driver/server.xml
@@ -33,7 +33,7 @@
 
     <dataSource id="fatDriver" jndiName="jdbc/fatDriver">
       <jdbcDriver libraryRef="FATDriverLib" internal.nonship.function="This is for internal development only. Never use this in production"/>
-      <properties url="jdbc:derby:memory:wrappedDerby;create=true" user="dbuser1" password="{xor}Oz0vKDtu" />
+      <properties url="jdbc:fatdriver:memory:jdbcdriver1;create=true" user="dbuser1" password="{xor}Oz0vKDtu" />
     </dataSource>
     
     <javaPermission codeBase="${server.config.dir}derby/FATDriver.jar" className="java.security.AllPermission"/>

--- a/dev/com.ibm.ws.jdbc_fat_driver/test-applications/fatdriver/src/jdbc/fat/driver/derby/FATDriver.java
+++ b/dev/com.ibm.ws.jdbc_fat_driver/test-applications/fatdriver/src/jdbc/fat/driver/derby/FATDriver.java
@@ -33,6 +33,11 @@ public class FATDriver implements Driver {
 
     @Override
     public Connection connect(String url, Properties info) throws SQLException {
+        if (!acceptsURL(url))
+            return null;
+
+        url = url.replace(":fatdriver:", ":derby:");
+
         System.out.println("[FATDriver]   ->  connect");
         try {
             Class.forName("org.apache.derby.jdbc.EmbeddedDriver");
@@ -51,7 +56,7 @@ public class FATDriver implements Driver {
 
     @Override
     public boolean acceptsURL(String url) throws SQLException {
-        throw new SQLFeatureNotSupportedException();
+        return url.startsWith("jdbc:fatdriver:");
     }
 
     @Override
@@ -61,7 +66,7 @@ public class FATDriver implements Driver {
 
     @Override
     public int getMajorVersion() {
-        return 0;
+        return 1;
     }
 
     @Override

--- a/dev/com.ibm.ws.jdbc_fat_driver/test-applications/jdbcapp/src/jdbc/fat/driver/web/JDBCDriverManagerServlet.java
+++ b/dev/com.ibm.ws.jdbc_fat_driver/test-applications/jdbcapp/src/jdbc/fat/driver/web/JDBCDriverManagerServlet.java
@@ -200,7 +200,7 @@ public class JDBCDriverManagerServlet extends FATServlet {
     @Test
     public void testFATDriver() throws Exception {
         Class.forName("jdbc.fat.driver.derby.FATDriver");
-        Connection conn = DriverManager.getConnection("jdbc:derby:memory:wrappedDerby;create=true");
+        Connection conn = DriverManager.getConnection("jdbc:fatdriver:memory:jdbcdriver1;create=true");
         try {
             System.out.println("Connected to " + conn.getMetaData().getDatabaseProductName());
             conn.createStatement().executeQuery("VALUES 1");

